### PR TITLE
Autogenerate GPU backend specific test files

### PR DESCRIPTION
**Pull request by masterleinad**
Based on top of #4268, related to https://github.com/kokkos/kokkos/pull/4268#discussion_r697730482.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

File: core/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -276,6 +276,43 @@ foreach(Tag Threads;Serial;OpenMP;Cuda;HPX;OpenMPTarget;HIP;SYCL)
   endif()
 endforeach()
 
+foreach(Tag HIP-HostPinned;Cuda-HostPinned;Cuda-UVM;SYCL-HostUSM;SYCL-SharedUSM)
```
Comment (dalg24) [inline core/unit_test/CMakeLists.txt:279]: `Tag` is not a good name here.  What about `PAIR_DEVICE_SPACE`?

File: core/unit_test/CMakeLists.txt [hunk 2]
```diff
@@ -493,27 +530,7 @@ if(Kokkos_ENABLE_CUDA)
...
-      cuda/TestCudaHostPinned_ViewAPI_b.cpp
-      cuda/TestCudaHostPinned_ViewAPI_c.cpp
-      cuda/TestCudaHostPinned_ViewAPI_d.cpp
-      cuda/TestCudaHostPinned_ViewAPI_e.cpp
-      cuda/TestCudaHostPinned_ViewCopy_a.cpp
-      cuda/TestCudaHostPinned_ViewCopy_b.cpp
-      cuda/TestCudaHostPinned_ViewMapping_a.cpp
-      cuda/TestCudaHostPinned_ViewMapping_b.cpp
-      cuda/TestCudaHostPinned_ViewMapping_subview.cpp
-      cuda/TestCudaUVM_SharedAlloc.cpp
```
Comment (dalg24) [inline core/unit_test/CMakeLists.txt:506]: What about `SharedAlloc`?

**Review by dalg24 — APPROVED** [review]
LGTM.  Would you please confirm at least for SYCL that when you list the tests in the executable you get the same count as in `develop`?

**Comment by masterleinad** [comment]
Should be good now.
With this pull request
```
10: Test command: /var/jenkins/workspace/Kokkos/build/core/unit_test/KokkosCore_UnitTest_SYCL3
10: Test timeout computed to be: 1500
10: [==========] Running 50 tests from 3 test cases.
10: [----------] Global test environment set-up.
10: [----------] 4 tests from sycl
10: [ RUN      ] sycl.team_scratch_1_queues
10: [       OK ] sycl.team_scratch_1_queues (54971 ms)
10: [ RUN      ] sycl.space_access
10: [       OK ] sycl.space_access (0 ms)
10: [ RUN      ] sycl.uvm
10: [       OK ] sycl.uvm (46 ms)
10: [ RUN      ] sycl.impl_view_accessible
10: [       OK ] sycl.impl_view_accessible (6 ms)
10: [----------] 4 tests from sycl (55023 ms total)
10: 
10: [----------] 23 tests from sycl_host_usm
10: [ RUN      ] sycl_host_usm.impl_shared_alloc
10: [       OK ] sycl_host_usm.impl_shared_alloc (47 ms)
10: [ RUN      ] sycl_host_usm.view_api_a
10: [       OK ] sycl_host_usm.view_api_a (11 ms)
10: [ RUN      ] sycl_host_usm.view_api_b
10: [       OK ] sycl_host_usm.view_api_b (5 ms)
10: [ RUN      ] sycl_host_usm.view_api_c
10: [       OK ] sycl_host_usm.view_api_c (1 ms)
10: [ RUN      ] sycl_host_usm.view_api_d
10: [       OK ] sycl_host_usm.view_api_d (6 ms)
10: [ RUN      ] sycl_host_usm.view_allocation_error
10: 
10: PI CUDA ERROR:
10: 	Value:           2
10: 	Name:            CUDA_ERROR_OUT_OF_MEMORY
10: 	Description:     out of memory
10: 	Function:        cuda_piextUSMHostAlloc
10: 	Source Location: /scratch/llvm/sycl/plugins/cuda/pi_cuda.cpp:4298
10: 
10: [       OK ] sycl_host_usm.view_allocation_error (0 ms)
10: [ RUN      ] sycl_host_usm.view_remap
10: [       OK ] sycl_host_usm.view_remap (2 ms)
10: [ RUN      ] sycl_host_usm.view_mirror_nonconst
10: [       OK ] sycl_host_usm.view_mirror_nonconst (1 ms)
10: [ RUN      ] sycl_host_usm.view_stride_method
10: [       OK ] sycl_host_usm.view_stride_method (0 ms)
10: [ RUN      ] sycl_host_usm.anonymous_space
10: [       OK ] sycl_host_usm.anonymous_space (0 ms)
10: [ RUN      ] sycl_host_usm.view_overload_resolution
10: [       OK ] sycl_host_usm.view_overload_resolution (2 ms)
10: [ RUN      ] sycl_host_usm.view_is_assignable
10: [       OK ] sycl_host_usm.view_is_assignable (25 ms)
10: [ RUN      ] sycl_host_usm.view_copy_tests
10: [       OK ] sycl_host_usm.view_copy_tests (10 ms)
10: [ RUN      ] sycl_host_usm.view_copy_tests_rank_0
10: [       OK ] sycl_host_usm.view_copy_tests_rank_0 (2 ms)
10: [ RUN      ] sycl_host_usm.view_copy_degenerated
10: [       OK ] sycl_host_usm.view_copy_degenerated (2 ms)
10: [ RUN      ] sycl_host_usm.view_mapping
10: [       OK ] sycl_host_usm.view_mapping (83 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_operator
10: [       OK ] sycl_host_usm.view_mapping_operator (4 ms)
10: [ RUN      ] sycl_host_usm.static_extent
10: [       OK ] sycl_host_usm.static_extent (0 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_atomic
10: [       OK ] sycl_host_usm.view_mapping_atomic (1 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_class_value
10: [       OK ] sycl_host_usm.view_mapping_class_value (0 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_assignable
10: [       OK ] sycl_host_usm.view_mapping_assignable (0 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_trivially_copyable
10: [       OK ] sycl_host_usm.view_mapping_trivially_copyable (0 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_subview
10: [       OK ] sycl_host_usm.view_mapping_subview (1 ms)
10: [----------] 23 tests from sycl_host_usm (203 ms total)
10: 
10: [----------] 23 tests from sycl_shared_usm
10: [ RUN      ] sycl_shared_usm.impl_shared_alloc
10: [       OK ] sycl_shared_usm.impl_shared_alloc (74 ms)
10: [ RUN      ] sycl_shared_usm.view_api_a
10: [       OK ] sycl_shared_usm.view_api_a (18 ms)
10: [ RUN      ] sycl_shared_usm.view_api_b
10: [       OK ] sycl_shared_usm.view_api_b (15 ms)
10: [ RUN      ] sycl_shared_usm.view_api_c
10: [       OK ] sycl_shared_usm.view_api_c (1 ms)
10: [ RUN      ] sycl_shared_usm.view_api_d
10: [       OK ] sycl_shared_usm.view_api_d (3 ms)
10: [ RUN      ] sycl_shared_usm.view_allocation_error
10: 
10: PI CUDA ERROR:
10: 	Value:           2
10: 	Name:            CUDA_ERROR_OUT_OF_MEMORY
10: 	Description:     out of memory
10: 	Function:        cuda_piextUSMSharedAlloc
10: 	Source Location: /scratch/llvm/sycl/plugins/cuda/pi_cuda.cpp:4347
10: 
10: [       OK ] sycl_shared_usm.view_allocation_error (0 ms)
10: [ RUN      ] sycl_shared_usm.view_remap
10: [       OK ] sycl_shared_usm.view_remap (2 ms)
10: [ RUN      ] sycl_shared_usm.view_mirror_nonconst
10: [       OK ] sycl_shared_usm.view_mirror_nonconst (0 ms)
10: [ RUN      ] sycl_shared_usm.view_stride_method
10: [       OK ] sycl_shared_usm.view_stride_method (0 ms)
10: [ RUN      ] sycl_shared_usm.anonymous_space
10: [       OK ] sycl_shared_usm.anonymous_space (0 ms)
10: [ RUN      ] sycl_shared_usm.view_overload_resolution
10: [       OK ] sycl_shared_usm.view_overload_resolution (1 ms)
10: [ RUN      ] sycl_shared_usm.view_is_assignable
10: [       OK ] sycl_shared_usm.view_is_assignable (6 ms)
10: [ RUN      ] sycl_shared_usm.view_copy_tests
10: [       OK ] sycl_shared_usm.view_copy_tests (26 ms)
10: [ RUN      ] sycl_shared_usm.view_copy_tests_rank_0
10: [       OK ] sycl_shared_usm.view_copy_tests_rank_0 (3 ms)
10: [ RUN      ] sycl_shared_usm.view_copy_degenerated
10: [       OK ] sycl_shared_usm.view_copy_degenerated (0 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping
10: [       OK ] sycl_shared_usm.view_mapping (72 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_operator
10: [       OK ] sycl_shared_usm.view_mapping_operator (15 ms)
10: [ RUN      ] sycl_shared_usm.static_extent
10: [       OK ] sycl_shared_usm.static_extent (0 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_atomic
10: [       OK ] sycl_shared_usm.view_mapping_atomic (1 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_class_value
10: [       OK ] sycl_shared_usm.view_mapping_class_value (1 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_assignable
10: [       OK ] sycl_shared_usm.view_mapping_assignable (0 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_trivially_copyable
10: [       OK ] sycl_shared_usm.view_mapping_trivially_copyable (0 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_subview
10: [       OK ] sycl_shared_usm.view_mapping_subview (2 ms)
10: [----------] 23 tests from sycl_shared_usm (240 ms total)
10: 
10: [----------] Global test environment tear-down
10: [==========] 50 tests from 3 test cases ran. (55466 ms total)
10: [  PASSED  ] 50 tests.
10/53 Test #10: KokkosCore_UnitTest_SYCL3 .....................   Passed   55.78 sec
```
With #4268 
```
10: Test command: /var/jenkins/workspace/Kokkos/build/core/unit_test/KokkosCore_UnitTest_SYCL3
10: Test timeout computed to be: 1500
10: [==========] Running 50 tests from 3 test cases.
10: [----------] Global test environment set-up.
10: [----------] 4 tests from sycl
10: [ RUN      ] sycl.team_scratch_1_queues
10: [       OK ] sycl.team_scratch_1_queues (52533 ms)
10: [ RUN      ] sycl.space_access
10: [       OK ] sycl.space_access (0 ms)
10: [ RUN      ] sycl.uvm
10: [       OK ] sycl.uvm (36 ms)
10: [ RUN      ] sycl.impl_view_accessible
10: [       OK ] sycl.impl_view_accessible (14 ms)
10: [----------] 4 tests from sycl (52583 ms total)
10: 
10: [----------] 23 tests from sycl_host_usm
10: [ RUN      ] sycl_host_usm.impl_shared_alloc
10: [       OK ] sycl_host_usm.impl_shared_alloc (41 ms)
10: [ RUN      ] sycl_host_usm.view_api_a
10: [       OK ] sycl_host_usm.view_api_a (7 ms)
10: [ RUN      ] sycl_host_usm.view_api_b
10: [       OK ] sycl_host_usm.view_api_b (7 ms)
10: [ RUN      ] sycl_host_usm.view_api_c
10: [       OK ] sycl_host_usm.view_api_c (1 ms)
10: [ RUN      ] sycl_host_usm.view_api_d
10: [       OK ] sycl_host_usm.view_api_d (9 ms)
10: [ RUN      ] sycl_host_usm.view_allocation_error
10: 
10: PI CUDA ERROR:
10: 	Value:           2
10: 	Name:            CUDA_ERROR_OUT_OF_MEMORY
10: 	Description:     out of memory
10: 	Function:        cuda_piextUSMHostAlloc
10: 	Source Location: /scratch/llvm/sycl/plugins/cuda/pi_cuda.cpp:4298
10: 
10: [       OK ] sycl_host_usm.view_allocation_error (0 ms)
10: [ RUN      ] sycl_host_usm.view_remap
10: [       OK ] sycl_host_usm.view_remap (1 ms)
10: [ RUN      ] sycl_host_usm.view_mirror_nonconst
10: [       OK ] sycl_host_usm.view_mirror_nonconst (2 ms)
10: [ RUN      ] sycl_host_usm.view_stride_method
10: [       OK ] sycl_host_usm.view_stride_method (0 ms)
10: [ RUN      ] sycl_host_usm.anonymous_space
10: [       OK ] sycl_host_usm.anonymous_space (0 ms)
10: [ RUN      ] sycl_host_usm.view_overload_resolution
10: [       OK ] sycl_host_usm.view_overload_resolution (1 ms)
10: [ RUN      ] sycl_host_usm.view_is_assignable
10: [       OK ] sycl_host_usm.view_is_assignable (37 ms)
10: [ RUN      ] sycl_host_usm.view_copy_tests
10: [       OK ] sycl_host_usm.view_copy_tests (6 ms)
10: [ RUN      ] sycl_host_usm.view_copy_tests_rank_0
10: [       OK ] sycl_host_usm.view_copy_tests_rank_0 (1 ms)
10: [ RUN      ] sycl_host_usm.view_copy_degenerated
10: [       OK ] sycl_host_usm.view_copy_degenerated (1 ms)
10: [ RUN      ] sycl_host_usm.view_mapping
10: [       OK ] sycl_host_usm.view_mapping (50 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_operator
10: [       OK ] sycl_host_usm.view_mapping_operator (2 ms)
10: [ RUN      ] sycl_host_usm.static_extent
10: [       OK ] sycl_host_usm.static_extent (0 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_atomic
10: [       OK ] sycl_host_usm.view_mapping_atomic (1 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_class_value
10: [       OK ] sycl_host_usm.view_mapping_class_value (0 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_assignable
10: [       OK ] sycl_host_usm.view_mapping_assignable (0 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_trivially_copyable
10: [       OK ] sycl_host_usm.view_mapping_trivially_copyable (0 ms)
10: [ RUN      ] sycl_host_usm.view_mapping_subview
10: [       OK ] sycl_host_usm.view_mapping_subview (1 ms)
10: [----------] 23 tests from sycl_host_usm (169 ms total)
10: 
10: [----------] 23 tests from sycl_shared_usm
10: [ RUN      ] sycl_shared_usm.impl_shared_alloc
10: [       OK ] sycl_shared_usm.impl_shared_alloc (45 ms)
10: [ RUN      ] sycl_shared_usm.view_api_a
10: [       OK ] sycl_shared_usm.view_api_a (19 ms)
10: [ RUN      ] sycl_shared_usm.view_api_b
10: [       OK ] sycl_shared_usm.view_api_b (13 ms)
10: [ RUN      ] sycl_shared_usm.view_api_c
10: [       OK ] sycl_shared_usm.view_api_c (1 ms)
10: [ RUN      ] sycl_shared_usm.view_api_d
10: [       OK ] sycl_shared_usm.view_api_d (4 ms)
10: [ RUN      ] sycl_shared_usm.view_allocation_error
10: 
10: PI CUDA ERROR:
10: 	Value:           2
10: 	Name:            CUDA_ERROR_OUT_OF_MEMORY
10: 	Description:     out of memory
10: 	Function:        cuda_piextUSMSharedAlloc
10: 	Source Location: /scratch/llvm/sycl/plugins/cuda/pi_cuda.cpp:4347
10: 
10: [       OK ] sycl_shared_usm.view_allocation_error (0 ms)
10: [ RUN      ] sycl_shared_usm.view_remap
10: [       OK ] sycl_shared_usm.view_remap (1 ms)
10: [ RUN      ] sycl_shared_usm.view_mirror_nonconst
10: [       OK ] sycl_shared_usm.view_mirror_nonconst (1 ms)
10: [ RUN      ] sycl_shared_usm.view_stride_method
10: [       OK ] sycl_shared_usm.view_stride_method (0 ms)
10: [ RUN      ] sycl_shared_usm.anonymous_space
10: [       OK ] sycl_shared_usm.anonymous_space (0 ms)
10: [ RUN      ] sycl_shared_usm.view_overload_resolution
10: [       OK ] sycl_shared_usm.view_overload_resolution (1 ms)
10: [ RUN      ] sycl_shared_usm.view_is_assignable
10: [       OK ] sycl_shared_usm.view_is_assignable (7 ms)
10: [ RUN      ] sycl_shared_usm.view_copy_tests
10: [       OK ] sycl_shared_usm.view_copy_tests (30 ms)
10: [ RUN      ] sycl_shared_usm.view_copy_tests_rank_0
10: [       OK ] sycl_shared_usm.view_copy_tests_rank_0 (1 ms)
10: [ RUN      ] sycl_shared_usm.view_copy_degenerated
10: [       OK ] sycl_shared_usm.view_copy_degenerated (0 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping
10: [       OK ] sycl_shared_usm.view_mapping (45 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_operator
10: [       OK ] sycl_shared_usm.view_mapping_operator (20 ms)
10: [ RUN      ] sycl_shared_usm.static_extent
10: [       OK ] sycl_shared_usm.static_extent (0 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_atomic
10: [       OK ] sycl_shared_usm.view_mapping_atomic (1 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_class_value
10: [       OK ] sycl_shared_usm.view_mapping_class_value (0 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_assignable
10: [       OK ] sycl_shared_usm.view_mapping_assignable (0 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_trivially_copyable
10: [       OK ] sycl_shared_usm.view_mapping_trivially_copyable (0 ms)
10: [ RUN      ] sycl_shared_usm.view_mapping_subview
10: [       OK ] sycl_shared_usm.view_mapping_subview (2 ms)
10: [----------] 23 tests from sycl_shared_usm (191 ms total)
10: 
10: [----------] Global test environment tear-down
10: [==========] 50 tests from 3 test cases ran. (52943 ms total)
10: [  PASSED  ] 50 tests.
10/53 Test #10: KokkosCore_UnitTest_SYCL3 .....................   Passed   57.85 sec
```
The diff only shows differences in run time.

**Review by Rombur — APPROVED** [review]

**Pull request merged**